### PR TITLE
Capture hot threads during slow cluster state application

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we emit a warning when applying the cluster state takes more than
30s by default. Experience shows that it's almost always one task that
exceeds the limit, and usually that task is the
`IndicesClusterStateService` (see #89821). This commit adds a `DEBUG`
logger that will capture a thread dump if a task is running for longer
than a configurable duration so we can get more insight into the cause
of the slow execution.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierRecordingService.java [hunk 1]
```diff
@@ -75,14 +87,39 @@ Releasable record(String action) {
...
+
+            if (logger.isDebugEnabled()) {
+                currentListener = new SubscribableListener<>();
+                currentListener.addTimeout(debugLoggingTimeout, threadPool, threadPool.generic());
+                currentListener.addListener(new ActionListener<>() {
+                    @Override
+                    public void onResponse(Void unused) {}
+
+                    @Override
+                    public void onFailure(Exception e) {
```
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierRecordingService.java:100]: nit: an assert that it is a timeout exception would also help understand the code better.

File: server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierRecordingServiceTests.java [hunk 2]
```diff
@@ -116,4 +135,42 @@ public void testRecordingServiceStats() {
         }
     }
 
+    @TestLogging(reason = "testing debug logging", value = "org.elasticsearch.cluster.service.ClusterApplierRecordingService:DEBUG")
+    public void testSlowTaskDebugLogging() {
+        final var debugLoggingTimeout = TimeValue.timeValueMillis(between(1, 100000));
+        var recorder = new Recorder(threadPool, debugLoggingTimeout);
```
Comment (pxsalehi) [inline server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierRecordingServiceTests.java:141]: maybe a tiny comment why these two cases in the test are different?

**Review by pxsalehi — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierRecordingService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierRecordingService.java:101]: ++ done in f9bd56f3dc55c3b7d562fecd1bf3e29b8850c0f6.

File: server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierRecordingServiceTests.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierRecordingServiceTests.java:141]: ++ done in f9bd56f3dc55c3b7d562fecd1bf3e29b8850c0f6.

Comment (pxsalehi) [reply server/src/test/java/org/elasticsearch/cluster/service/ClusterApplierRecordingServiceTests.java:141]: Thank you, it just prevents squinting too long at them! :)))

**Pull request merged**